### DISCUSSION TITLE
Reroute back to original route as soon as possible

### DISCUSSION THREAD

**Discussion by mkrezdo** [devemux86/cruiser#497] [Q&A] [answered]
Hello all,

sorry for a perhaps simple question, but for me it is not possible to test it out at the moment, so I hope someone can answer following question:

I plan my route on a pc and want to navigate it 100% like it is planned (for that I use normally the kurviger format / sometimes gpx tracks).
Because of this I deactivated the reroute/recalculation in the Cruiser settings. I had bad experience with several navigation solutions in the past, that after rerouting the complete route was diffrent.

So my target & the question: Is it possible to use a setting that - in case of missing the route (eg because of a blocked street) - trys to go back to the original rout as soon as possible but without changing anything behind the point when I got back on the original route?

Hope I could describe it clear enough :-)

Thx in advance,
Marco

**Comment by devemux86** [comment]
Hi @mkrezdo,

> Is it possible to use a setting that - in case of missing the route (eg because of a blocked street) - trys to go back to the original rout as soon as possible but without changing anything behind the point when I got back on the original route?

As you described, this is exactly how rerouting already works in Cruiser navigation:
- It calculates a new re-route to the next point (via or shaping)

- The rest of the original route remains **unchanged**

So you can be sure that you navigate your route as it was planned. :slightly_smiling_face: 

If you've seen something else, can you post a route example so we can check it?

**Reply by mkrezdo** [reply]
I guess then it does not exactly what I mean. You said it will go back to the next point as soon as possible. But if I've a route with not many points there can be big changes after the reroute. 

I try to make an exampel: a 150 km route ( https://kurv.gr/XthLB ) planned in kurviger with max. curvy profile. I can import this 1:1. 
There are no additional points in the route (because they are not needed)
But if there is a road blocked after the first 10 km - is Cruiser than really try to reroute only this area or will it change the complete route to the highway because it is faster?

![image](https://github.com/devemux86/cruiser/assets/137488443/c89290b5-fe7f-42ef-bfd7-4116ba4e7e20)

![image](https://github.com/devemux86/cruiser/assets/137488443/5a975e18-2cb0-4eee-aa91-9082a721fbd5)

**Reply by devemux86** [reply]
@mkrezdo thanks for the extra details!

Rerouting now happens to the next waypoint (via or shaping).

>  But if I've a route with not many points there can be big changes after the reroute.
>  There are no additional points in the route (because they are not needed)

An alternative (we had at the beginning) is rerouting back to the route, regardless of waypoints.
If there is interest, I can restore it, probably as option?

So if there are not many waypoints (via or shaping), at least it will get back on route very soon.

> But if there is a road blocked after the first 10 km - is Cruiser than really try to reroute only this area or will it change the complete route to the highway because it is faster?

In blocked areas you can use the "Avoid roadblock" function (tap the top-left direction icon).

The rerouting depends on the selected routing options.
Have you selected BRouter (offline) and avoid motorways and other main roads in routing settings?

Next Cruiser will also have:
- https://github.com/devemux86/cruiser/discussions/254

**Reply by 0709wiwiwi** [reply]
With back to track or trackpoints, I personally have not had good experiences.
Recalculation to a shaping point is clear plus is selectable and easy to skip.
Cruiser can already add shaping points without a recalculation trigger.

**Reply by devemux86** [reply]
> With back to track or trackpoints, I personally have not had good experiences.
Recalculation to a shaping point is clear (selectable and skipable)

That's why we had removed it and preferred rerouting to next shaping / via point.
But it can be useful in cases like the above route 150 km without extra waypoints.

> Cruiser can already add shaping points without a recalculation trigger.

Now in Beta, it will be available in the next version:
- https://github.com/devemux86/cruiser/discussions/482

**Reply by mkrezdo** [reply]
> An alternative (we had at the beginning) is rerouting back to the route, regardless of waypoints. If there is interest, I can restore it, probably as option?
> 
> So if there are not many waypoints (via or shaping), at least it will get back on route very soon.

That sound's like what I'm looking for. I understand that someone does not like it because it is undefined. But I prefere this.
I such an case I zoom out and drive along the road, the redirection and check the map as long as the way back looks good for me.

> In blocked areas you can use the "Avoid roadblock" function (tap the top-left direction icon).

But this is not possible without stopping the bike and doing sth on the phone. This is a reason why I do not like it that much.
It does also happens that I only missed a road :-) - also then this will not really help

> The rerouting depends on the selected routing options. Have you selected BRouter (offline) and avoid motorways and other main roads in routing settings?
> 
> Next Cruiser will also have:
> 
> * [Curvy, extra curvy, custom routes with BRouter (offline) #254](https://github.com/devemux86/cruiser/discussions/254)

Yes I've choosen BRouter offline and I already installed the Beta. This new feature is great as backup on the road. But also this will not create the same route - it's a bit diffrent than how Kurviger is planning

**Comment by 0709wiwiwi** [comment]
@mkrezdo 

> Now in Beta, it will be available in the next version.

**Reply by mkrezdo** [reply]
could be a good workaround. But same here. For that you've to stop the bike and do s.th on the phone.

Zoom out a bit to orientate is possible during driving (using Barbuttons)

**Reply by 0709wiwiwi** [reply]
Positioning a shaping point not causing any recalculation must be done very precisely **indeed.**

**Reply by devemux86** [reply]
> Positioning a shaping point not causing any recalculation must be done very precisely indeed.

It is actually very simple:
- Long press route and set point (shaping or via)
- Press OK without moving the map

**Reply by devemux86** [reply]
Add shaping points on a route is an action to be done in route planning e.g. at home.
Not while driving, where everything should be handled by the app.

I will examine how we can offer the old rerouting without waypoints.

**Reply by devemux86** [reply]
As an extra option.

**Reply by 0709wiwiwi** [reply]
> Not while driving, where everything should be handled by the app.

When faced with a barrage with a local diversion you probably stop for a moment and decide what to do.

**Reply by WolfgangZ42** [reply]
Willi,
because I'm a lazy guy, I do this by [gpxshaping.de](url) and get even more shaping points: 
<img width="500" src="https://github.com/devemux86/cruiser/assets/121806238/3f239607-7fc7-4ad0-bf24-8362414bb5a0"/>
But nevertheless - setting enough shaping points is the right direction!

**Comment by mkrezdo** [comment]
additional Test on my way home from work yesterday. I created a Kurviger route extra curvy. 
Then after a few 100 m I drove left on the highway instead of straight on as the rout was planned. (green arrow)
Best case I could expect is that it'll take me back on the blue planned route as soon as possible (pink arrow)
![image](https://github.com/devemux86/cruiser/assets/137488443/1417848b-b0d2-41b8-aa4f-532cb3f01533)

But what it did is that. Replan the red route (light green was the original). In Cruiser BRouter offline "fastest route" is choosen.
Regardless of that I do not want to drive the complete different red route - why is it planning this also very curvy route an not the fastest one over the highway?
![image](https://github.com/devemux86/cruiser/assets/137488443/e2678d04-ed8b-4160-b5e1-41f3ccab3780)

**Reply by devemux86** [reply]
@mkrezdo thanks for the examples!

As explained above, rerouting now calculates a new re-route to the next waypoint (shaping or via).
So the route should have several shaping points to not change in rerouting.

_We will add a new rerouting option that will work regardless of route's waypoints._

> In Cruiser BRouter offline "fastest route" is choosen.

What do you mean?
If not want fastest route, then select a different route profile that avoid motorways and main roads:
- https://github.com/devemux86/cruiser/discussions/254

**Comment by devemux86** [comment]
I can also add a function in Cruiser that adds automatically shaping points in a route.

Should this distance be in settings or fixed, e.g. every 10 km?

**Reply by bios4** [reply]
> I can also add a function in Cruiser that adds automatically shaping points in a route.

Do you mean that Cruiser adds shaping points always to a route, or just in case a planned route is left or the "avoid roadblock" option was selected?

**Reply by 0709wiwiwi** [reply]
Perfect.
I did use 10km in the manual edited [example](https://github.com/devemux86/cruiser/discussions/497#discussioncomment-9126254) it is easy to count. 
Best in Settings ? FREE.

> or just in case a planned route is left or the "avoid roadblock" option was selected?

Good question.

**Reply by devemux86** [reply]
> Do you mean that Cruiser adds shaping points always to a route

Not always, it will be a user action in route menu "Add shaping points",
like the other route actions that can be called only by the users.

>  just in case a planned route is left or the "avoid roadblock" option was selected?

In rerouting, Cruiser can handle this in the background without extra points. :slightly_smiling_face: 
We can add a rerouting setting to use waypoints or not, as mentioned above.

---

The 2 things can exist together with each other:
- A rerouting setting seems to be needed if user decides not using many waypoints
- "Add shaping points" is for convenience if user wants to do it in an automatic way

**Reply by bios4** [reply]
> A rerouting setting seems to be needed if user decides not using many waypoints

Would this the be something like "when the route is left, reroute back to the initially planned route withing the next "X" kilometers" (where "X" is a setting)?

> "Add shaping points" is for convenience if user wants to do it in an automatic way

Would this be to be selected in the long-press route menu, basically to "set shaping points to the loaded route every "X" kilometers" (where "X" is a setting, too)?

The second one could be achieved with a "popup" with a slider, where one can select the wanted distance of the shaping points (similar when selecting "Show on map" in "My collection").

**Reply by devemux86** [reply]
> Would this the be something like "when the route is left, reroute back to the initially planned route withing the next "X" kilometers" (where "X" is a setting)?

Do we really need a distance setting for this? Just return to the route at the nearest / next location.

> Would this be to be selected in the long-press route menu, basically to "set shaping points to the loaded route every "X" kilometers" (where "X" is a setting, too)?

Correctly, like this.

**Reply by bios4** [reply]
> Do we really need a distance setting for this? Just return to the route at the nearest / next location.

Fair point...

And when I do not drive back on the route at the nearest / next location, Cruiser will recalculate again to the "then nearest" possibility!?

> Correctly, like this.

Quick and dirty scetch (with this you could avoid a separate "global" [per routing profile] setting, which might be insuffiecient for routes of different lengths at all):

<img src="https://github.com/devemux86/cruiser/assets/145460147/028c4762-7e13-4491-bd6f-1b30acd4f74e" width="50%">

Provide slider values in different non-linear distances (e.g. 1, 2, 5, 10, 15 kilometers)

**Reply by WolfgangZ42** [reply]
> I can also add a function in Cruiser that adds automatically shaping points in a route.

I'd appreciate that! 
Maybe it could be done only when 

- avoid road block or
- re-routing is initiated (automatically or manually)

and maybe it's enough to set just one shaping point (the next one only)?

**Reply by devemux86** [reply]
> And when I do not drive back on the route at the nearest / next location, Cruiser will recalculate again to the "then nearest" possibility!?

When being off route and if the new setting to not use the waypoints in rerouting is selected.

About the distance, this can be in app settings to avoid too many dialogs.
Also it should be different per profile, as e.g. walk has different distances.

**Reply by devemux86** [reply]
@WolfgangZ42 please see my above [comment](https://github.com/devemux86/cruiser/discussions/497#discussioncomment-9127436).

> In rerouting, Cruiser can handle this in the background without extra points. 🙂
We can add a rerouting setting to use waypoints or not.

**Reply by bios4** [reply]
> About the distance, this can be in app settings

Please see my updated comment above...

I could also have routes of different lengths in the same profile!???
Some of my trips are only 15 kilometers total, others are several hundred...

So ONE setting per profile would NOT fit for all of my routes, which means I would have to re-set it repeatedly.

If you would implement the slider solution, with a "pre-stored" (set it once [per profile], keep it as long as not changed) interval (like done for the zoom levels in "My collection"), this would be a one-time setting for all users who do not need different intervals, but would offer a simple possibility to adapt the setting individually for others who are in the need of it.

**Reply by devemux86** [reply]
@bios4 thanks for the details.

> I could also have routes of different lengths in the same profile

This is also true, so we would need the distance in the dialog.
The check box is not needed, just select distance and press OK (or cancel).

**Reply by bios4** [reply]
> The check box is not needed, just select distance and press OK (or cancel).

Altough off-topic here: this then also applies to the "Show on map" menus in "My collection":

<img src="https://github.com/devemux86/cruiser/assets/145460147/c7de6040-b264-45b0-ad7a-16047458f5e1" width="33%">

Maybe you want to adapt these also then to have a common approach for these menus in the app!??
Set the slider to the most left to 0 (= don't show on map), or set any zoom level when items should be displayed.

**Reply by devemux86** [reply]
They are different, the favorites dialog only changes settings with 2 options.

Shaping points dialog makes an action with 1 option.

**Reply by devemux86** [reply]
> Set the slider to the most left to 0 (= don't show on map)

We might use this in some settings, because Android settings have not very rich interface,
unless we write them ourselves, as I did with the color settings.

Here this is a user dialog, we can construct it as we like, so better be more understandable.

**Reply by bios4** [reply]
> Shaping points dialog makes an action with 1 option.

Do you plan for providing an "undo" option for this action to remove the added shaping points again?

**Reply by devemux86** [reply]
> Do you plan for providing an "undo" option for this action to remove the added shaping points again?

Cruiser has a history of the route changes, so undo / redo are always available. :slightly_smiling_face:

**Reply by bios4** [reply]
> Cruiser has a history of the route changes, so undo / redo are always available.

You're talking about the round arrow- and X-buttons in the middle of the screen, right?
How can I get them back after (accidentially) closing them with tapping the "X"?

**Reply by devemux86** [reply]
>  How can I get them back after (accidentially) closing them with tapping the "X"?

Long press on the map and they are at the bottom of the menu dialog.

**Reply by bios4** [reply]
Puh - that these arrows there represent the very same functions as the on-screen buttons were not self-explaining to me!
Thanks for clarifying...

**Comment by devemux86** [comment]
1. Rerouting to next waypoint (current default)

3. Rerouting back to original route (available via a new option)

Note that the direction arrow also changes, showing the direction to the next waypoint or just the route.

|1. Rerouting to next waypoint|
|-|
|<img src="https://github.com/devemux86/cruiser/assets/3484020/ff5e51b9-776f-42c9-be7a-f60fb9ed1f2a" width="50%"/><img src="https://github.com/devemux86/cruiser/assets/3484020/ec8b4135-ee35-4518-b513-a18774c6c152" width="50%"/>|

|2. Rerouting back to original route|
|-|
|<img src="https://github.com/devemux86/cruiser/assets/3484020/775e49ed-fba7-4947-9549-9f3e8944599f" width="50%"/><img src="https://github.com/devemux86/cruiser/assets/3484020/0fe15bb1-480b-4614-8eca-aac565d2bbf2" width="50%"/>|

**Reply by bios4** [reply]
This looks very promising!

> Note that the direction arrow also changes, showing the direction to the next waypoint or just the route.

Could there be an easy way to somehow show if the direction arrow points "just back to the route" or to a specified waypoint?
I'm not sure myself how this could look like or if this is even useful - just free floating thoughts...

**Reply by devemux86** [reply]
The new setting (2nd row), we can change / improve the text.

<img src="https://github.com/devemux86/cruiser/assets/3484020/7564c391-f82c-4510-95c2-4e6b6db48c79" width="50%"/>

**Reply by bios4** [reply]
Suggestion - rename "Recalculate route with waypoints" to "Recalculate route TO NEXT waypoint".

If the toggle is off, this would then translate to something like "Recalculate route to nearest point ahead"!?
Can this be then shown as a description?

If this is not possible, maybe it is better to explicitly offer these two options as radiobuttons?
To be self-explaining...

**Reply by devemux86** [reply]
@bios4 thanks, I will change the texts.

**Reply by devemux86** [reply]
Do you think this new mode should be the default, as it reroutes quickly back to route?

Or keep as default the current mode that reroutes to next waypoint?

**Reply by mkrezdo** [reply]
> Suggestion - rename "Recalculate route with waypoints" to "Recalculate route TO NEXT waypoint".

I'm of the same opinion

> Do you think this new mode should be the default, as it reroutes quickly back to route?
> 
> Or keep as default the current mode that reroutes to next waypoint?

For me this does not matter - to have to option the change it would be great.
I cannot imagine what people prefere more...but often it's better not to change the standard behavior. If a user is updating the app he could be surprised why the re-routing works diffrent. So perhaps it's better not to set it on default

**Reply by devemux86** [reply]
> but often it's better not to change the standard behavior. If a user is updating the app he could be surprised why the re-routing works diffrent. So perhaps it's better not to set it on default

@mkrezdo thanks, this is a good point! :+1:

**Reply by bios4** [reply]
> @bios4 thanks, I will change the texts.

If you keep the toggle (instead of radiobuttons with both options explained), I would appreciate if the description would contain the different options somehow, e.g.:

<img src="https://github.com/devemux86/cruiser/assets/145460147/7a8cb043-e104-4adc-b0ef-34d21f1314c4" width="33%">

Optionally, you maybe could combine the first two options from this ^^ screenshot in **ONE setting** with radiobuttons!??

- O Recalculation off
- O Recalculate to next waypoint
- O Recalculate to nearest point ahead

**Reply by devemux86** [reply]
Better go for the radio buttons:

**Rerouting mode**
- Recalculate route to next waypoint
- Recalculate route to nearest point

**Reply by devemux86** [reply]
Combined with the unique "**Strict route following**" in Cruiser using all points or just via points,
we can have a very interesting and _personalized_ navigation.

<img src="https://github.com/devemux86/cruiser/assets/3484020/503416fe-cc2c-49f1-b9d3-715175524aaf" width="50%"/>

**Reply by bios4** [reply]
Suggestion - change the second option to "Recalculate route to nearest point _of original route_" (should fit nicely in terms of screen space)

**Reply by devemux86** [reply]
@bios4 thanks, we can do this.

**Comment by devemux86** [comment]
The new rerouting mode is available for testing in the latest:
- https://github.com/devemux86/cruiser/discussions/482

(Settings | Navigation | Recalculate route)

**Comment by 0709wiwiwi** [comment]
I have tested this virtually now anyway.  ONLY by a simulation !
It produces the same nonsense as before with the kurviger app.

**Reply by devemux86** [reply]
@0709wiwiwi thanks for the report.

This new rerouting mode does not add extra shaping points.
Everything is handled internally and only the original route waypoints are used.

Rerouting tries to return to the nearest next location after exiting the route.
If you are far enough ahead, the nearest next location would be clearly further ahead.

It all depends on the geometry of the route and the location of the user.
Please post specific route examples and user location to check.

**Reply by 0709wiwiwi** [reply]
Even through a Lockito app simulation (speed 50kmh), it takes some time to test, but I did notice that a shaping point 1 was added to the original design with only a start and a stop.   I sure can forward you the trajectory.

**Reply by devemux86** [reply]
@0709wiwiwi thanks for the example.

This can currently happen when the route does not have via / shaping points.
Otherwise extra shaping points are not added. I will fix this case too.

**Comment by airald** [comment]
I tested this feature going home. I make a route with some diversions I will not take to force cruiser to reroute. I see strange things like that, never seen before.

![image](https://github.com/devemux86/cruiser/assets/20151719/545153b6-9a71-4eb2-8c4c-95b62da55b17)
![image](https://github.com/devemux86/cruiser/assets/20151719/f8062214-651c-41bf-b561-32348b38afd9)

When Cruiser reroute and makes me go to the shaping point that was skipped for me, I select "skip next waypoint" and this happens.

**Reply by devemux86** [reply]
@airald thanks for the report.

I cannot understand the route and the rerouting from the images. :slightly_smiling_face: 

Such complicated routes with overlapping parts better use the default rerouting mode with waypoints.

The new rerouting mode is more useful in normal routes and only works in rerouting.
The other cases, like skip waypoints, avoid road blocks, all have their own reroutings.

**Reply by airald** [reply]
The problem is there are straight lines where there are not any street or path

![image](https://github.com/devemux86/cruiser/assets/20151719/e07f2151-62d6-49e6-a45e-9f38a22d47dc)

This example there route is over the train station, you can see in this image

![image](https://github.com/devemux86/cruiser/assets/20151719/3e912bd2-02c3-4aa6-8b90-15128de0b961)

It's surprising that there's not any waypoint here.

![image](https://github.com/devemux86/cruiser/assets/20151719/37e2ee30-1ca4-4532-9b92-e4211d586ff1)

**Reply by mkrezdo** [reply]
I had a chance to do a first short test this morning too. Also I had my problems with the new option.
I've no screenshot available but there where two points that were not good after I missed the route:
- after missing the route Cruiser recalculated it. When this was done the "start point" was set on the position were I missed the route. So the new route tried to turn and go back to this point to come back on the the original route. But it would be much faster to follow the actual street a few more meters and then leave it to go on a new way back to the nearest point of the old route
- seconed problem was that - after I ignored the request to drive back the rerouting stopped. It seems that it will only calculate a new rout once. If you miss the new route then again it will not do anything more.

**Reply by 0709wiwiwi** [reply]
Are you sure you are using the latest beta b10 ? 
https://github.com/devemux86/cruiser/discussions/497#discussioncomment-9159363

**Reply by mkrezdo** [reply]
No I stilll hava the b8, but I'd no time to test before :-)

**Reply by devemux86** [reply]
@mkrezdo thanks for the report.

For routes without many points, better use the new tool to automatically add shaping points.
(available in the route menu in the latest Beta)

Then the rerouting will reroute to the next point and the _remaining route will not change_.

**Comment by devemux86** [comment, marked as answer]
I think the most safe solution is to not add this new rerouting mode.
Instead use the current rerouting with waypoints which has been stable all this time.

And with the new tool of adding automatically extra shaping points, there will be no problem:

- https://github.com/devemux86/cruiser/discussions/499

**Reply by airald** [reply]
I always use many waypoints in my routes and If I use reroute it works good (usally I return to route by myself), I just was testing this new option.
